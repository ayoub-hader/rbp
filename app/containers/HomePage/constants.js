export const FILL_FORM = 'app/HomePage/FILL_FORM';
