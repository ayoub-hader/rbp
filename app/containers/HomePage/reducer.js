import produce from 'immer';
import {
  CLEAN_UP,
  SET_USER,
  SET_USER_ERROR,
  SET_USER_SUCCESS,
} from './constants';

export const initialState = {
  user: {},
  loading: false,
  success: false,
  error: null,
};

const homePageReducer = produce((draft, action) => {
  switch (action.type) {
    case SET_USER:
      draft.user = action.user;
      draft.success = false;
      draft.loading = true;
      draft.error = null;
      break;
    case SET_USER_SUCCESS:
      draft.success = true;
      draft.loading = false;
      draft.error = null;
      break;
    case SET_USER_ERROR:
      draft.success = false;
      draft.loading = false;
      draft.error = action.error;
      break;
    case CLEAN_UP:
      draft.user = {};
      draft.loading = false;
      draft.success = false;
      draft.error = null;
  }
}, initialState);

export default homePageReducer;
