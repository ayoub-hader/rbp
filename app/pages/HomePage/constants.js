export const SET_USER = 'app/HomePage/SET_USER';
export const SET_USER_SUCCESS = 'app/HomePage/SET_USER_SUCCESS';
export const SET_USER_ERROR = 'app/HomePage/SET_USER_ERROR';
export const CLEAN_UP = 'app/HomePage/CLEAN_UP';
