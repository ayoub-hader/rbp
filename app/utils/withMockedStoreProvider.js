import React from 'react';
import { Provider } from 'react-redux';
import PropTypes from 'prop-types';
import storeMock from 'utils/storeMock';
import { DEFAULT_LOCALE } from 'locales';
import { IntlProvider } from 'react-intl';

const withMockedStoreProvider = (
  Component,
  store = storeMock({}),
  locale = DEFAULT_LOCALE,
) => {
  const TestComponent = props => (
    <Provider store={store}>
      <IntlProvider locale={locale}>
        <Component {...props} />
      </IntlProvider>
    </Provider>
  );

  TestComponent.propTypes = {
    store: PropTypes.object,
    locale: PropTypes.object,
  };
  return TestComponent;
};

export default withMockedStoreProvider;
