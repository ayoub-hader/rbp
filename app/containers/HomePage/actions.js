import { FILL_FORM } from './constants';

export function fillFormAction(params) {
  return {
    type: FILL_FORM,
    params,
  };
}
