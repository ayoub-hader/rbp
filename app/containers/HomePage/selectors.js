import { createSelector } from 'reselect';
import { initialState } from './reducer';

const selectHomePageDomain = state => state.homePage || initialState;
const makeSelectForm = createSelector(
  selectHomePageDomain,
  substate => substate.form,
);

const makeSelectHomePage = () =>
  createSelector(selectHomePageDomain, substate => substate);

export default makeSelectHomePage;
export { selectHomePageDomain, makeSelectForm };
