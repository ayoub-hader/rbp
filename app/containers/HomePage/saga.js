import { call, put, takeLatest } from 'redux-saga/effects';
import request from 'utils/request';
import { setUserErrorAction, setUserSuccessAction } from './actions';
import { SET_USER } from './constants';

export function* saveUser({ user }) {
  try {
    //const { data } = yield call(request.post, "endpoint");
    console.log('im in saga');
    yield put(setUserSuccessAction());
  } catch (error) {
    yield put(setUserErrorAction(error));
  }
}

export default function* homePageSaga() {
  yield takeLatest(SET_USER, saveUser);
}
