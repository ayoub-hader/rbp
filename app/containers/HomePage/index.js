import React from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { useInjectReducer, useInjectSaga } from 'redux-injectors';
import { useForm } from 'react-hook-form';
import * as yup from 'yup';
import { yupResolver } from '@hookform/resolvers/yup';
import registerYupLocale from 'utils/yupValidator';
import Button from 'components/Button';
import FormInput from 'components/FormInput';
import { setUserAction } from './actions';
import reducer from './reducer';
import {
  makeSelectUser,
  makeSelectUserError,
  makeSelectUserSuccess,
} from './selectors';
import saga from './saga';

registerYupLocale();

const stateSelector = createStructuredSelector({
  user: makeSelectUser,
  success: makeSelectUserSuccess,
  error: makeSelectUserError,
});

const SignupSchema = yup.object().shape({
  firstName: yup.string().required(),
  email: yup.string().required().email(),
  age: yup.number().required().positive().integer().min(15).max(20),
});

function HomePage() {
  useInjectReducer({ key: 'homePage', reducer });
  useInjectSaga({ key: 'homePage', saga });

  const dispatch = useDispatch();
  const { user, success, error } = useSelector(stateSelector);

  const {
    register,
    handleSubmit,
    formState: { errors },
  } = useForm({
    resolver: yupResolver(SignupSchema),
  });

  const onSubmit = (data) => {
    dispatch(setUserAction(data));
  };

  return (
    <form onSubmit={handleSubmit(onSubmit)}>
      {user ? <div>Current user name : {user?.firstName}</div> : null}
      <div>
        <FormInput
          label="firstName"
          type="texte"
          register={register}
          errors={errors.firstName}
        />
      </div>
      <div>
        <FormInput
          label="email"
          type="texte"
          register={register}
          errors={errors.email}
        />
      </div>
      <div>
        <FormInput
          label="age"
          type="number"
          register={register}
          errors={errors.age}
        />
      </div>
      <Button value={'submit'} />
    </form>
  );
}

export default HomePage;
