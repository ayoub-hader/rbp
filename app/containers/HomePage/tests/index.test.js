import React from 'react';
import { render } from '@testing-library/react';
import HomePage from 'containers/HomePage/index';
import withMockedStoreProvider from 'utils/withMockedStoreProvider';

const HomePageContainer = withMockedStoreProvider(HomePage);

describe('<HomePage />', () => {
  it('should render and match the snapshot', () => {
    const {
      container: { firstChild },
    } = render(<HomePageContainer />);
    expect(firstChild).toMatchSnapshot();
  });
});
