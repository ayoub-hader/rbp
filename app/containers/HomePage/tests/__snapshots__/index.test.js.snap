// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HomePage /> should render and match the snapshot 1`] = `
<form>
  <div>
    Current user name : 
  </div>
  <div>
    <label>
      firstName
    </label>
    <input
      name="firstName"
      type="texte"
    />
  </div>
  <div>
    <label>
      email
    </label>
    <input
      name="email"
      type="texte"
    />
  </div>
  <div>
    <label>
      age
    </label>
    <input
      name="age"
      type="number"
    />
  </div>
  <input
    type="submit"
  />
</form>
`;
